File: src/commands/fun/balance.ts
import { SlashCommandBuilder, EmbedBuilder, ChatInputCommandInteraction } from 'discord.js';
import User from '../../models/User.js';

export default {
  data: new SlashCommandBuilder()
    .setName('balance')
    .setDescription('Check your coin balance and XP')
    .addUserOption(option =>
      option.setName('target').setDescription('User to check (defaults to yourself)').setRequired(false)
    ),
  async execute(interaction: ChatInputCommandInteraction): Promise<void> {
    const user = interaction.options.getUser('target') || interaction.user;

    let profile;
    try {
      profile = await User.findOne({ discordId: user.id }).lean();
    } catch (_) {
      await interaction.reply({ content: 'Database is unavailable.', ephemeral: true });
      return;
    }

    if (!profile) {
      const isSelf = user.id === interaction.user.id;
      await interaction.reply({
        content: isSelf
          ? 'You have no profile yet. Use `/daily` or `/work` to get started!'
          : `${user.tag} has no profile yet.`,
        ephemeral: true
      });
      return;
    }

    const embed = new EmbedBuilder()
      .setAuthor({ name: user.tag, iconURL: user.displayAvatarURL({ size: 64 }) })
      .setColor('#f59e0b')
      .addFields(
        { name: '💰 Coins', value: profile.balance.toLocaleString(), inline: true },
        { name: '⭐ XP', value: profile.xp.toLocaleString(), inline: true },
        { name: '🏆 Level', value: profile.level.toString(), inline: true }
      );

    await interaction.reply({ embeds: [embed], ephemeral: true });
  }
};

File: src/bot/commands/fun/balance.rs
use crate::bot::{Context, Error};
use crate::db::models::User;
use poise::serenity_prelude as serenity;

/// Check your coin balance and XP
#[poise::command(slash_command, ephemeral)]
pub async fn balance(
    ctx: Context<'_>,
    #[description = "User to check (defaults to yourself)"] target: Option<serenity::User>,
) -> Result<(), Error> {
    let user = target.as_ref().unwrap_or_else(|| ctx.author());

    let Some(db) = ctx.data().database() else {
        ctx.say("Database is unavailable.").await?;
        return Ok(());
    };

    let profile = User::find_by_discord_id(&db, &user.id.to_string()).await?;

    match profile {
        None => {
            let is_self = user.id == ctx.author().id;
            let msg = if is_self {
                "You have no profile yet. Use `/daily` or `/work` to get started!".to_string()
            } else {
                format!("{} has no profile yet.", user.name)
            };
            ctx.say(msg).await?;
        }
        Some(p) => {
            let embed = serenity::CreateEmbed::new()
                .author(serenity::CreateEmbedAuthor::new(user.name.clone()).icon_url(user.face()))
                .colour(serenity::Colour::from_rgb(245, 158, 11))
                .field("💰 Coins", p.balance.to_string(), true)
                .field("⭐ XP", p.xp.to_string(), true)
                .field("🏆 Level", p.level.to_string(), true);

            ctx.send(poise::CreateReply::default().embed(embed)).await?;
        }
    }
    Ok(())
}
